File: C_plus_plus/Symmetric_Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode *root) {
        if(!root) return true;
        return recursive(root->left,root->right);
    }
    bool recursive(TreeNode *left,TreeNode *right){
        if(!left&&!right) return true;
        if(!left||!right||left->val!=right->val) return false;
        return recursive(left->left,right->right)&&recursive(left->right,right->left);
    }
};

File: java/Symmetric_Tree.java
// Recursive solution
// Time is O(n), space is O(n) since using recursive.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        return isSym(root.left,root.right);
    }
    public boolean isSym(TreeNode a, TreeNode b) {
        if (a == null && b == null) return true;
        if (a == null || b == null) return false;
        if (a.val != b.val) return false;
        return isSym(a.left, b.right) && isSym(a.right,b.left);
    }
}
